File: specs/json-schema/research/string_probe/Len.java
// Java half of the string-constraint probe (see main.go for the contract).
// Run: java Len.java   (JDK 11+ single-file mode)
import java.util.regex.*;

public class Len {
  public static void main(String[] a) {
    String s = "a😀b"; // a + 😀 (surrogate pair) + b
    // LENGTH: String.length() counts UTF-16 code UNITS — wrong for astral chars.
    // codePointCount(0, length()) is the spec-correct code-point count.
    System.out.println("JAVA length(UTF-16 units): " + s.length()
        + "  codePointCount(codepoints): " + s.codePointCount(0, s.length())); // 4, 3
    // No normalization: NFC (1 code point) vs NFD (2). All four agree per form.
    String nfc = "\u00e9", nfd = "e\u0301";
    System.out.println("JAVA NFC u00e9 codePointCount: " + nfc.codePointCount(0, nfc.length()));   // 1
    System.out.println("JAVA NFD e+u0301 codePointCount: " + nfd.codePointCount(0, nfd.length())); // 2

    // PATTERN: use Matcher.find() (unanchored) — NOT matches() (anchors the whole
    // input, a silent footgun). Use DEFAULT flags: `.` is already code-point-aware
    // (matches astral), and \d\w\s stay ASCII (ECMA-262-aligned). Do NOT set
    // UNICODE_CHARACTER_CLASS — it flips \d\w\s to Unicode and diverges.
    System.out.println("JAVA 'a.b' find 'a😀b' (. = code point): " + Pattern.compile("a.b").matcher(s).find()); // true
    System.out.println("JAVA 'cat' find    'the cat sat' (unanchored): " + Pattern.compile("cat").matcher("the cat sat").find());
    System.out.println("JAVA 'cat' matches 'the cat sat' (anchored footgun): " + Pattern.compile("cat").matcher("the cat sat").matches()); // false
    System.out.println("JAVA '\\d' find Arabic digit ٣ (DEFAULT): " + Pattern.compile("\\d").matcher("٣").find()); // false (ASCII)
    System.out.println("JAVA '\\d' find Arabic digit ٣ (UNICODE_CHARACTER_CLASS): "
        + Pattern.compile("\\d", Pattern.UNICODE_CHARACTER_CLASS).matcher("٣").find()); // true <- do NOT use
    System.out.println("JAVA '\\w' find u00e9 (DEFAULT): " + Pattern.compile("\\w").matcher("\u00e9").find()); // false (ASCII)
    // Perl features Java accepts but RE2 rejects (gated out at load):
    System.out.println("JAVA lookahead compiles: " + tryCompile("(?=foo)"));
    System.out.println("JAVA backref compiles: " + tryCompile("(a)\\1"));
  }

  static boolean tryCompile(String p) {
    try { Pattern.compile(p); return true; } catch (PatternSyntaxException e) { return false; }
  }
}

File: specs/json-schema/research/string_probe/len.mjs
// JS half of the string-constraint probe (see main.go for the contract).
// LENGTH: .length counts UTF-16 code *units* — wrong for astral chars.
// [...s].length spreads by code point — the spec-correct count.
const s = "a\u{1F600}b";
console.log("JS   .length(UTF-16 units):", s.length, " [...s].length(codepoints):", [...s].length); // 4, 3
// No normalization: NFC é (1 code point) vs NFD é (2). All four agree per form.
console.log("JS   NFC u00e9 [...].length:", [...("\u00e9")].length);    // 1
console.log("JS   NFD e+u0301 [...].length:", [...("e\u0301")].length); // 2

// PATTERN: .test() is unanchored. The `u` flag is MANDATORY for P1 — it makes
// `.` and quantifiers code-point-aware (matching Go RE2 / Python / Java), while
// JS `\d\w\s` stay ASCII (ECMA-262). Without `u`, `.` is a UTF-16 unit and
// diverges on astral input.
console.log("JS   /a.b/  test 'a😀b' (NO u flag):", /a.b/.test(s));   // false  <- divergent
console.log("JS   /a.b/u test 'a😀b' (u flag):", /a.b/u.test(s));     // true   <- aligned
console.log("JS   /cat/u test 'the cat sat' (unanchored):", /cat/u.test("the cat sat"));
console.log("JS   /\\d/u matches Arabic digit ٣:", /\d/u.test("٣"));  // false (ASCII)
console.log("JS   /\\w/u matches u00e9:", /\w/u.test("\u00e9"));             // false (ASCII)
// RE2-rejected Perl features that JS accepts (why RE2 is the load-time gate):
console.log("JS   lookahead compiles:", (() => { try { new RegExp("(?=foo)", "u"); return true; } catch { return false; } })());
console.log("JS   backref compiles:", (() => { try { new RegExp("(a)\\1", "u"); return true; } catch { return false; } })());
